### PR TITLE
feat(core, apps): 'PacketDataUnmarshaler' interface added and implemented

### PR DESCRIPTION
**Pull request by srdtrk**
## Description

This PR reduces the diff when reviewing the adr8 PR https://github.com/cosmos/ibc-go/pull/3939.

`PacketDataUnmarshaler` is an optional interface which allows a middleware to request the packet data to be unmarshaled by the base application.

closes: #XXXX

### Commit Message / Changelog Entry

```text
feat(core, apps): 'PacketDataUnmarshaler' interface added and implemented
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: testing/mock/mock.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ const (
 var (
 	MockAcknowledgement             = channeltypes.NewResultAcknowledgement([]byte("mock acknowledgement"))
 	MockFailAcknowledgement         = channeltypes.NewErrorAcknowledgement(fmt.Errorf("mock failed acknowledgement"))
+	ErrorMock                       = fmt.Errorf("mock failed acknowledgement")
```
Comment (colin-axner) [inline testing/mock/mock.go:38]: On channel upgrades, we added an error to test application callbacks with. Maybe I can upstream this change and we can use that? see https://github.com/cosmos/ibc-go/blob/04-channel-upgrades/testing/mock/mock.go#L42

Comment (colin-axner) [reply testing/mock/mock.go:38]: see #4189

**Review by colin-axner — APPROVED** [review]
Nice one! Only request would be to use the mock application error once #4189 is merged

**Review by crodriguezvega — APPROVED** [review]

File: modules/apps/29-fee/ibc_middleware_test.go [hunk 2]
```diff
@@ -1080,3 +1083,26 @@ func (suite *FeeTestSuite) TestGetAppVersion() {
...
+	suite.Require().NoError(err)
+	suite.Require().Equal(ibcmock.MockPacketData, packetData)
+}
+
+func (suite *FeeTestSuite) TestPacketDataUnmarshalerInterfaceError() {
+	// test the case when the underlying application cannot be casted to a PacketDataUnmarshaler
+	mockFeeMiddleware := fee.NewIBCMiddleware(nil, feekeeper.Keeper{})
+
+	_, err := mockFeeMiddleware.UnmarshalPacketData(ibcmock.MockPacketData)
+	suite.Require().ErrorIs(err, errorsmod.Wrapf(types.ErrUnsupportedAction, "underlying app does not implement %T", (*porttypes.PacketDataUnmarshaler)(nil)))
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/29-fee/ibc_middleware_test.go]: done

File: modules/apps/transfer/ibc_module_test.go [hunk 3]
```diff
@@ -238,3 +242,59 @@ func (suite *TransferTestSuite) TestOnChanOpenAck() {
...
 }
+
+func (suite *TransferTestSuite) TestPacketDataUnmarshalerInterface() {
+	var (
+		sender           = sdk.AccAddress(secp256k1.GenPrivKey().PubKey().Address()).String()
+		receiver         = sdk.AccAddress(secp256k1.GenPrivKey().PubKey().Address()).String()
+		srcCallbackAddr  = sdk.AccAddress(secp256k1.GenPrivKey().PubKey().Address()).String()
+		destCallbackAddr = sdk.AccAddress(secp256k1.GenPrivKey().PubKey().Address()).String()
+		denom            = "transfer/channel-0/atom"
+		amount           = "100"
```
[the remark this answers is not in the document]
Comment (srdtrk) [reply modules/apps/transfer/ibc_module_test.go]: I've improved this by removing callbacks, adding a new test case, and using test coin

File: testing/mock/mock.go (hunk 1 above)
Comment (srdtrk) [reply testing/mock/mock.go]: using this error now

**Pull request merged**